File: tests/workflow/producer/rust/src/stat.rs
use console::style;
use indicatif::{ProgressBar, ProgressStyle};
use std::collections::HashMap;

mod expectations {
    pub const BeaconA: usize = 1;
    pub const BeaconsBeaconA: usize = 1;
    pub const BeaconsBeaconB: usize = 1;
    pub const BeaconsSubBeaconA: usize = 1;
    pub const GroupAStructA: usize = 1;
    pub const GroupAStructB: usize = 1;
    pub const GroupBGroupCStructA: usize = 1;
    pub const GroupBGroupCStructB: usize = 2;
    pub const GroupBStructA: usize = 1;
    pub const StructA: usize = 1;
    pub const StructB: usize = 2;
    pub const StructC: usize = 2;
    pub const StructD: usize = 3;
    pub const StructE: usize = 3;
    pub const StructF: usize = 2;
    pub const StructJ: usize = 2;
    pub const TriggerBeacons: usize = 1;
    pub const FinishConsumerTestBroadcast: usize = 1;
    pub const Connected: usize = 1;
    pub const Disconnected: usize = 1;
    pub const Shutdown: usize = 1;
    pub const StructEmpty: usize = 2;
    pub const StructEmptyA: usize = 1;
    pub const StructEmptyB: usize = 1;
}

#[derive(PartialEq, Hash, Clone)]
pub enum Alias {
    BeaconA,
    BeaconsBeaconA,
    BeaconsBeaconB,
    BeaconsSubBeaconA,
    GroupAStructA,
    GroupAStructB,
    GroupBGroupCStructA,
    GroupBGroupCStructB,
    GroupBStructA,
    StructA,
    StructB,
    StructC,
    StructD,
    StructE,
    StructF,
    StructJ,
    TriggerBeacons,
    FinishConsumerTestBroadcast,
    Connected,
    Disconnected,
    Shutdown,
    StructEmpty,
    StructEmptyA,
    StructEmptyB,
}

impl Eq for Alias {}

impl std::fmt::Display for Alias {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::BeaconA => write!(f, "BeaconA"),
            Self::BeaconsBeaconA => write!(f, "BeaconsBeaconA"),
            Self::BeaconsBeaconB => write!(f, "BeaconsBeaconB"),
            Self::BeaconsSubBeaconA => write!(f, "BeaconsSubBeaconA"),
            Self::GroupAStructA => write!(f, "GroupAStructA"),
            Self::GroupAStructB => write!(f, "GroupAStructB"),
            Self::GroupBGroupCStructA => write!(f, "GroupBGroupCStructA"),
            Self::GroupBGroupCStructB => write!(f, "GroupBGroupCStructB"),
            Self::GroupBStructA => write!(f, "GroupBStructA"),
            Self::StructA => write!(f, "StructA"),
            Self::StructB => write!(f, "StructB"),
            Self::StructC => write!(f, "StructC"),
            Self::StructD => write!(f, "StructD"),
            Self::StructE => write!(f, "StructE"),
            Self::StructF => write!(f, "StructF"),
            Self::StructJ => write!(f, "StructJ"),
            Self::TriggerBeacons => write!(f, "TriggerBeacons"),
            Self::FinishConsumerTestBroadcast => write!(f, "FinishConsumerTestBroadcast"),
            Self::Connected => write!(f, "Connected"),
            Self::Disconnected => write!(f, "Disconnected"),
            Self::Shutdown => write!(f, "Shutdown"),
            Self::StructEmpty => write!(f, "StructEmpty"),
            Self::StructEmptyA => write!(f, "StructEmptyA"),
            Self::StructEmptyB => write!(f, "StructEmptyB"),
        }
    }
}

pub enum StatEvent {
    Inc(Alias),
}

pub struct Stat {
    done: usize,
    pub tests: HashMap<Alias, (usize, usize)>,
    pub indexes: HashMap<Alias, usize>,
}

impl Stat {
    pub fn new() -> Self {
        let mut tests = HashMap::new();
        tests.insert(Alias::BeaconA, (0, expectations::BeaconA));
        tests.insert(Alias::BeaconsBeaconA, (0, expectations::BeaconsBeaconA));
        tests.insert(Alias::BeaconsBeaconB, (0, expectations::BeaconsBeaconB));
        tests.insert(
            Alias::BeaconsSubBeaconA,
            (0, expectations::BeaconsSubBeaconA),
        );
        tests.insert(Alias::GroupAStructA, (0, expectations::GroupAStructA));
        tests.insert(Alias::GroupAStructB, (0, expectations::GroupAStructB));
        tests.insert(
            Alias::GroupBGroupCStructA,
            (0, expectations::GroupBGroupCStructA),
        );
        tests.insert(
            Alias::GroupBGroupCStructB,
            (0, expectations::GroupBGroupCStructB),
        );
        tests.insert(Alias::GroupBStructA, (0, expectations::GroupBStructA));
        tests.insert(Alias::StructA, (0, expectations::StructA));
        tests.insert(Alias::StructB, (0, expectations::StructB));
        tests.insert(Alias::StructC, (0, expectations::StructC));
        tests.insert(Alias::StructD, (0, expectations::StructD));
        tests.insert(Alias::StructE, (0, expectations::StructE));
        tests.insert(Alias::StructF, (0, expectations::StructF));
        tests.insert(Alias::StructJ, (0, expectations::StructJ));
        tests.insert(Alias::TriggerBeacons, (0, expectations::TriggerBeacons));
        tests.insert(
            Alias::FinishConsumerTestBroadcast,
            (0, expectations::FinishConsumerTestBroadcast),
        );
        tests.insert(Alias::Connected, (0, expectations::Connected));
        tests.insert(Alias::Disconnected, (0, expectations::Disconnected));
        tests.insert(Alias::Shutdown, (0, expectations::Shutdown));
        tests.insert(Alias::StructEmpty, (0, expectations::StructEmpty));
        tests.insert(Alias::StructEmptyA, (0, expectations::StructEmptyA));
        tests.insert(Alias::StructEmptyB, (0, expectations::StructEmptyB));
        Self {
            tests,
            done: 0,
            indexes: HashMap::new(),
        }
    }

    pub fn apply(&mut self, event: StatEvent) {
        match event {
            StatEvent::Inc(alias) => {
                if let Some((current, _)) = self.tests.get_mut(&alias) {
                    *current += 1;
                }
            }
        }
    }

    pub fn get_index(&mut self, alias: Alias) -> usize {
        *self.indexes.entry(alias.clone()).or_insert(0) += 1;
        self.indexes[&alias]
    }

    pub fn expectation(&self) -> usize {
        let mut all: usize = 0;
        for (_alias, (_current, expectation)) in &self.tests {
            all += expectation;
        }
        all
    }

    pub fn get_beacons_count(&self) -> usize {
        self.tests[&Alias::BeaconA].0
            + self.tests[&Alias::BeaconsBeaconA].0
            + self.tests[&Alias::BeaconsBeaconB].0
            + self.tests[&Alias::BeaconsSubBeaconA].0
    }
}

impl std::fmt::Display for Stat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut output = format!("{}\n", "=".repeat(70));
        let mut tests = vec![];
        for (alias, (current, expectation)) in &self.tests {
            tests.push((alias.to_string(), current, expectation));
        }
        tests.sort();
        for (alias, current, expectation) in &tests {
            let mut alias = alias.to_string();
            let filler = 60 - alias.len();
            alias = format!(
                "{}{}",
                alias,
                ".".repeat(if filler > 0 { filler } else { 0 })
            );
            output = format!(
                "{} {} {}: {}/{}\n",
                output,
                if current == expectation {
                    style("☑").bold().green().dim()
                } else {
                    style("☐").bold().red().dim()
                },
                style(format!("[{}]", alias)).bold().dim(),
                if current == expectation {
                    style(format!("{}", current)).dim()
                } else {
                    style(format!("{}", current)).bold().red().dim()
                },
                if current == expectation {
                    style(format!("{}", expectation)).dim()
                } else {
                    style(format!("{}", expectation)).bold().red().dim()
                }
            );
        }
        output = format!("{}{}\n", output, "=".repeat(70));
        write!(f, "{}", output)
    }
}

File: tests/workflow/producer/typescript/src/stat.ts
export const expectations = {
	BeaconA: 1,
	BeaconsBeaconA: 1,
	BeaconsBeaconB: 1,
	BeaconsSubBeaconA: 1,
	GroupAStructA: 4,
	GroupAStructB: 4,
	GroupBGroupCStructA: 3,
	GroupBGroupCStructB: 5,
	GroupBStructA: 4,
	StructA: 5,
	StructB: 2,
	StructC: 6,
	StructD: 2,
	StructE: 3,
	StructF: 2,
	StructJ: 2,
	TriggerBeacons: 1,
	FinishConsumerTestBroadcast: 1,
	Connected: 1,
	Disconnected: 1,
	StructEmpty: 2,
	StructEmptyA: 1,
	StructEmptyB: 1,
	Error: 0,
};

export enum Alias {
	BeaconA = "BeaconA",
	BeaconsBeaconA = "BeaconsBeaconA",
	BeaconsBeaconB = "BeaconsBeaconB",
	BeaconsSubBeaconA = "BeaconsSubBeaconA",
	GroupAStructA = "GroupAStructA",
	GroupAStructB = "GroupAStructB",
	GroupBGroupCStructA = "GroupBGroupCStructA",
	GroupBGroupCStructB = "GroupBGroupCStructB",
	GroupBStructA = "GroupBStructA",
	StructA = "StructA",
	StructB = "StructB",
	StructC = "StructC",
	StructD = "StructD",
	StructF = "StructF",
	StructJ = "StructJ",
	StructE = "StructE",
	TriggerBeacons = "TriggerBeacons",
	FinishConsumerTestBroadcast = "FinishConsumerTestBroadcast",
	Connected = "Connected",
	Disconnected = "Disconnected",
	StructEmpty = "StructEmpty",
	StructEmptyA = "StructEmptyA",
	StructEmptyB = "StructEmptyB",
	Error = "Error",
}

const all: Alias[] = [
	Alias.BeaconA,
	Alias.BeaconsBeaconA,
	Alias.BeaconsBeaconB,
	Alias.BeaconsSubBeaconA,
	Alias.GroupAStructA,
	Alias.GroupAStructB,
	Alias.GroupBGroupCStructA,
	Alias.GroupBGroupCStructB,
	Alias.GroupBStructA,
	Alias.StructA,
	Alias.StructB,
	Alias.StructC,
	Alias.StructD,
	Alias.StructE,
	Alias.StructF,
	Alias.StructJ,
	Alias.TriggerBeacons,
	Alias.FinishConsumerTestBroadcast,
	Alias.Connected,
	Alias.Disconnected,
	Alias.StructEmpty,
	Alias.StructEmptyA,
	Alias.StructEmptyB,
	Alias.Error,
];

export interface TestData {
	done: number;
	expectation: number;
}

export class Stat {
	static MIN_BEACONS_COUNT: number = 4;
	private _tests: Map<Alias, TestData> = new Map();
	private _indexes: Map<Alias, number> = new Map();

	constructor(noExpectations?: boolean) {
		all.forEach((alias) => {
			this._indexes.set(alias, 0);
		});
		this._tests.set(Alias.BeaconA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconA,
		});
		this._tests.set(Alias.BeaconsBeaconA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconsBeaconA,
		});
		this._tests.set(Alias.BeaconsBeaconB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconsBeaconB,
		});
		this._tests.set(Alias.BeaconsSubBeaconA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.BeaconsSubBeaconA,
		});
		this._tests.set(Alias.GroupAStructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupAStructA,
		});
		this._tests.set(Alias.GroupAStructB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupAStructB,
		});
		this._tests.set(Alias.GroupBGroupCStructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupBGroupCStructA,
		});
		this._tests.set(Alias.GroupBGroupCStructB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupBGroupCStructB,
		});
		this._tests.set(Alias.GroupBStructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.GroupBStructA,
		});
		this._tests.set(Alias.StructA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructA,
		});
		this._tests.set(Alias.StructB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructB,
		});
		this._tests.set(Alias.StructC, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructC,
		});
		this._tests.set(Alias.StructD, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructD,
		});
		this._tests.set(Alias.StructE, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructE,
		});
		this._tests.set(Alias.StructF, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructF,
		});
		this._tests.set(Alias.StructJ, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructJ,
		});
		this._tests.set(Alias.TriggerBeacons, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.TriggerBeacons,
		});
		this._tests.set(Alias.FinishConsumerTestBroadcast, {
			done: 0,
			expectation: noExpectations
				? 0
				: expectations.FinishConsumerTestBroadcast,
		});
		this._tests.set(Alias.Connected, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.Connected,
		});
		this._tests.set(Alias.Disconnected, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.Disconnected,
		});
		this._tests.set(Alias.StructEmpty, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructEmpty,
		});
		this._tests.set(Alias.StructEmptyA, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructEmptyA,
		});
		this._tests.set(Alias.StructEmptyB, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.StructEmptyB,
		});
		this._tests.set(Alias.Error, {
			done: 0,
			expectation: noExpectations ? 0 : expectations.Error,
		});
	}

	public index(alias: Alias): number {
		const index = this._indexes.get(alias) as number;
		this._indexes.set(alias, index + 1);
		return index;
	}

	public case(alias: Alias) {
		const data = this._tests.get(alias) as TestData;
		data.done += 1;
		this._tests.set(alias, data);
	}

	public getBeaconsCount(): number {
		return (
			(this._tests.get(Alias.BeaconA) as TestData).done +
			(this._tests.get(Alias.BeaconsBeaconA) as TestData).done +
			(this._tests.get(Alias.BeaconsBeaconB) as TestData).done +
			(this._tests.get(Alias.BeaconsSubBeaconA) as TestData).done
		);
	}

	public merge(stat: Stat) {
		this._tests.forEach((data: TestData, alias: Alias) => {
			const income = stat.getCase(alias);
			data.done += income.done;
			data.expectation += income.expectation;
		});
	}

	public print() {
		const LEN: number = 70;
		this._tests.forEach((data: TestData, alias: Alias) => {
			const title = `${alias}`;
			const correct = data.done === data.expectation;
			console.log(
				`${correct ? "☑" : "☐"} ${title}${".".repeat(
					LEN - title.length
				)}: ${data.done} / ${data.expectation}`
			);
		});
	}

	public getCase(alias: Alias): TestData {
		return this._tests.get(alias) as TestData;
	}

	public getErrors(): string[] {
		const errors: string[] = [];
		this._tests.forEach((data: TestData, alias: Alias) => {
			if (data.done !== data.expectation) {
				errors.push(
					`Test for usecase "${alias}" is failed; ${data.done} / ${data.expectation}`
				);
			}
		});
		return errors;
	}
}
